File: NextcloudTalk/Calls/AvatarBackgroundImageView.m
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "AvatarBackgroundImageView.h"

@implementation GradientView

@dynamic layer;

+ (Class)layerClass {
    return [CAGradientLayer class];
}

@end

@implementation AvatarBackgroundImageView

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self initGradientLayer];
    }
    
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self initGradientLayer];
    }
    
    return self;
}

- (instancetype)initWithImage:(UIImage *)image
{
    self = [super initWithImage:image];
    if (self) {
        [self initGradientLayer];
    }
    
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self initGradientLayer];
    }
    
    return self;
}

- (void)initGradientLayer
{
    _gradientView = [[GradientView alloc] initWithFrame:self.bounds];
    _gradientView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _gradientView.layer.colors = @[(id)[[UIColor colorWithWhite:0 alpha:0.6] CGColor], (id)[[UIColor colorWithWhite:0 alpha:0.6] CGColor]];
    _gradientView.layer.locations = @[@0.0, @1.0];
    [self addSubview:_gradientView];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
//    _gradientLayer.frame = self.bounds;
}

@end

File: NextcloudTalk/Calls/AvatarBackgroundImageView.swift
//
// SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

import UIKit

class GradientView: UIView {

    override class var layerClass: AnyClass {
        return CAGradientLayer.self
    }

    // Swift can't narrow the inherited `layer` type the way the ObjC @dynamic declaration did
    var gradientLayer: CAGradientLayer {
        // swiftlint:disable:next force_cast
        return self.layer as! CAGradientLayer
    }
}

@objcMembers class AvatarBackgroundImageView: UIImageView {

    var gradientView: GradientView?

    override init(frame: CGRect) {
        super.init(frame: frame)
        initGradientLayer()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        initGradientLayer()
    }

    private func initGradientLayer() {
        let gradientView = GradientView(frame: self.bounds)
        gradientView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        gradientView.gradientLayer.colors = [UIColor(white: 0, alpha: 0.6).cgColor,
                                             UIColor(white: 0, alpha: 0.6).cgColor]
        gradientView.gradientLayer.locations = [0.0, 1.0]

        self.addSubview(gradientView)
        self.gradientView = gradientView
    }
}
